### PR TITLE
Limit PFB IDs to 254 or fewer characters (#3313)

### PR DESCRIPTION
**Pull request by jessebrennan**
https://github.com/DataBiosphere/azul/issues/3313

Author

- [x] PR title references issue
- [x] PR title matches issue title (preceded by `Fix: ` for bugs)   <sub>or there is a good reason why they're different</sub>
- [x] Title of main commit references issue
- [x] PR is connected to Zenhub issue and description links to issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>
- [x] Added announcement to PR description                  <sub>or this PR does not require announcement</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [x] Commented in issue about demo expectations            <sub>or labelled issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear              <sub>or issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                               <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [x] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [ ] Made announcement requested by author                 <sub>or PR description does not contain an announcement</sub>
- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing or does not target `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing or does not target `prod`</sub>

Operator

- [ ] Unassigned PR

### CODE REVIEW

File: src/azul/service/avro_pfb.py
Comment (amarjandu) [inline src/azul/service/avro_pfb.py:153]: nit. IMO its better to place assertions for instance attributes within `def __attrs_post_init__(self):` if for some reason this evolves, or a different class constructor is used you still get the benefits of the assertion.

This idea goes against the idea of asserting attributes before creating an instance w/ them in order to avoid malformed instances floating around in the runtime, but I've yet to think of a way that these malformed instances can exist without the runtime exiting due to the assertion/requirement firing off...

**Review by amarjandu — APPROVED** [review]
When you recreated the pfb results file, it looks like the `file_source` field moved around, I'm guessing the old canned result had some type of non-determinism that was solved between now and then?

Other than that just a nit.

File: src/azul/service/avro_pfb.py
Comment (jessebrennan) [reply src/azul/service/avro_pfb.py:153]: This is a good suggestion, and inspired me to use attrs validators.

**Comment by amarjandu** [comment]
```
Index: src/azul/service/avro_pfb.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/service/avro_pfb.py b/src/azul/service/avro_pfb.py
--- a/src/azul/service/avro_pfb.py	(revision 59330ae32450ef90a35de2d3fb6f0e710d1f6e0c)
+++ b/src/azul/service/avro_pfb.py	(date 1631033776758)
@@ -125,17 +125,17 @@
             yield entity.to_json(sorted(relations, key=attrgetter('dst_name', 'dst_id')))
 
 
-def validate_id(instance, attribute, value):
-    require(len(value) <= 254, 'Terra requires IDs be no longer than 254 chars')
-
-
 @attr.s(auto_attribs=True, frozen=True, kw_only=True)
 class PFBEntity:
     """
     Python representation of the PFB data object. Attribute names conform to the
     PFB spec (which simplifies serialization).
     """
-    id: str = attr.ib(validator=validate_id)
+
+    def _validate_id(self, attribute, value):
+        require(len(value) <= 254, 'Terra requires IDs be no longer than 254 chars')
+        
+    id: str = attr.ib(validator=_validate_id)
     name: str
     object: MutableJSON = attr.ib(eq=False)
     namespace_uuid: ClassVar[UUID] = UUID('bc93372b-9218-4f0e-b64e-6f3b339687d6')
```

from our conversation.

**Review by amarjandu — CHANGES_REQUESTED** [review]
Per our discussion, attempt to scope `validate_id` within the class to keep the class portable.

**Comment by jessebrennan** [comment]
> validate_id within the class to keep the class portable.

I tried, but making it a class method doesn't work, unfortunately.

**Review by amarjandu — APPROVED** [review]

File: src/azul/service/avro_pfb.py
Comment (hannes-ucsc) [inline src/azul/service/avro_pfb.py:136]: Use `__attrs_post_init__`.

File: src/azul/service/avro_pfb.py
Comment (hannes-ucsc) [inline src/azul/service/avro_pfb.py:157]: Assert that joiner doesn't occur in input.

```suggestion
        id_ = f"{name}_{uuid5(cls.namespace_uuid, '_'.join(ids))}_{len(ids}}"
```

File: src/azul/service/avro_pfb.py
Comment (hannes-ucsc) [reply src/azul/service/avro_pfb.py:136]:
```suggestion
    def _validate_id(self, attribute, value):
        reject(len(value) > 254, 'Terra requires IDs be no longer than 254 chars', len(value))
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: src/azul/service/avro_pfb.py
Comment (jessebrennan) [reply src/azul/service/avro_pfb.py:136]: Instead of logging `len(value)`, I think logging just `value` would be more helpful.

File: src/azul/service/avro_pfb.py
Comment (jessebrennan) [reply src/azul/service/avro_pfb.py:157]: > Assert that joiner doesn't occur in input.
If I apply this to `name` too, then my diff looks different from your suggestion, so I'm intentionally diverging.

File: src/azul/service/avro_pfb.py
Comment (hannes-ucsc) [inline src/azul/service/avro_pfb.py:158]: Q

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: src/azul/service/avro_pfb.py
Comment (hannes-ucsc) [inline src/azul/service/avro_pfb.py:158]: Assert that `name` does not contain underscore.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: src/azul/service/avro_pfb.py
Comment (jessebrennan) [reply src/azul/service/avro_pfb.py:158]: @hannes-ucsc: "Since name may contain an underscore, use `:` or `/` for the second joiner. Also use .join instead of format string."

**Comment by jessebrennan** [comment]
@hannes-ucsc  I'm unable to test the `.` in the ID field in Terra from my personal deployment (see https://humancellatlas.slack.com/archives/C018YPVKC73/p1631667541000100?thread_ts=1627329947.003500&cid=C018YPVKC73), at least until this change is merged into `dev`.

We can either
- merge this anyway and hope nothing breaks
- substitute `.` for something else and slightly less reverseable
- wait for Terra to whitelist Azul's personal deployments.

This is blocked until we decide on an approach above.

File: src/azul/strings.py
Comment (hannes-ucsc) [inline src/azul/strings.py:157]: Let's not make this a library function. If we did, I'd have a few more concerns that I don't want to spend time on now.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
I can't write demo instructions without repro. Please add one to ticket.

**Comment by jessebrennan** [comment]
Reproduction added to description.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
The reason we're doing this is because Terra rejected our manifest so a reproduction must involve Terra.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Conflicts. Reproduction is lacking expected and actual outcomes.

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**
